File: p161m/one_edit_distance.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isOneEditDistance = function(s, t) {
  if (s.length < t.length) [s, t] = [t, s];
  if (s.length - t.length > 1) return false;
  let editedOnce = false;
  for (let i = 0, j = 0; j < t.length; ++i, ++j) {
    if (s[i] === t[j]) continue;
    if (editedOnce) return false;
    editedOnce = true;
    if (s.length > t.length) j--;
  }
  return editedOnce || s.length > t.length;
};
// TEST
[
  ['ab', 'acdd', false],
  ['ab', 'acd', false],
  ['cab', 'ad', false],
  ['cat', 'cut', true],
  ['cat', 'cast', true],
  ['cat', 'cats', true],
  ['cat', 'at', true],
  ['cat', 'dog', false],
  ['cat', 'act', false],
  ['cat', 'cats', true]
].forEach(t => {
  const actual = isOneEditDistance(t[0], t[1]);
  console.log('Is', t[0], t[1], 'one edit distance? ->', actual);
  console.assert(actual === t[2]);
});

File: p161m/one_edit_distance.py
class Solution:
    def isOneEditDistance(self, s: str, t: str) -> bool:
        if len(s) < len(t):
            s, t = t, s
        if len(s) - len(t) > 1:
            return False
        i, j, edit_once = 0, 0, False
        while j < len(t):
            if s[i] != t[j]:
                if edit_once:
                    return False
                edit_once = True
                if len(s) > len(t):
                    j = j - 1
            i, j = i + 1, j + 1
        return edit_once or len(s) > len(t)


# TESTS
tests = [
    ["ab", "acdd", False],
    ["ab", "acd", False],
    ["cab", "ad", False],
    ["cat", "cut", True],
    ["cat", "cast", True],
    ["cat", "cats", True],
    ["cat", "at", True],
    ["cat", "dog", False],
    ["cat", "act", False],
    ["cat", "cats", True],
    ["cat", "cat", False],
]
for t in tests:
    sol = Solution()
    actual = sol.isOneEditDistance(t[0], t[1])
    print("Is", t[0], t[1], "one edit distance? ->", actual)
    assert actual == t[2]
